jax/_src/numpy/lax_numpy.py

```diff
@@ -1593,8 +1593,9 @@ def tile(A, reps):
                for rep in reps)
   A_shape = (1,) * (len(reps) - ndim(A)) + shape(A)
   reps = (1,) * (len(A_shape) - len(reps)) + reps
-  result = broadcast_to(reshape(A, [j for i in A_shape for j in [1, i]]),
-                        [k for pair in zip(reps, A_shape) for k in pair])
+  target_shape = tuple(k for r, s in zip(reps, A_shape) for k in ((r * s,) if r == 1 or s == 1 else (r, s)))
+  broadcast_dims = tuple(np.cumsum([1 if r == 1 or s == 1 else 2 for r, s in zip(reps, A_shape)]) - 1)
+  result = lax.broadcast_in_dim(A, target_shape, broadcast_dims[-ndim(A):])
```

Comment:
Maybe something like the following will make this code more obvious
```suggestion
  rs_pairs = [(r * s,) if r == 1 or s == 1 else (r, s) for r, s in zip(reps, A_shape)]
  target_shape = tuple(chain.from_iterable(rs_pairs))
  broadcast_dims = tuple(np.cumsum(list(map(len, rs_pairs))) - 1)[-ndim(A):]
  result = lax.broadcast_in_dim(A, target_shape, broadcast_dims)
```

Reply:
Great catch, and your suggestion make sense! What do you think @jakevdp ?

Reply:
I really don't like the double list comprehension, even `np.concatenate` looks better.

Reply:
Ah, I agree nested-for comprehension is not pretty. 
But it is faster than `np.concatenate`, and doesn't need `from itertools import chain`. 
```python
x = jnp.ones((1,) * 100 + (2,) * 10)
print(timeit('jax.make_jaxpr(lambda a: tile(a, 2))(x)', globals=globals(), number=100))
# 5% faster than np.concatenate
```
Original implementation use nested-for comprehension too :P.